lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py

```diff
@@ -338,49 +334,101 @@ def run_once(heartbeat_handler: Any, younger_than: int, nattempts: int, vos: Opt
-                                                                              recoverable_replicas[vo][rse_key][replica_key]["scope"],
-                                                                              recoverable_replicas[vo][rse_key][replica_key]["name"],
-                                                                              nattempts)
-                                    for reason in suspicious_reason:
-                                        if "auditor" in reason["reason"].lower():
-                                            auditor += 1
-                                            files_to_be_declared_bad.append(recoverable_replicas[vo][rse_key][replica_key])
-                                            break
-                                        elif "checksum" in reason["reason"].lower():
-                                            checksum += 1
-                                            files_to_be_declared_bad.append(recoverable_replicas[vo][rse_key][replica_key])
-                                            break
-                                else:
-                                    logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Match for the metadata 'datatype' (%s) of replica found in json file, but no match for 'action' (%s)",
-                                           rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'], i["datatype"], i["action"])
-                                break
-                            else:
-                                # If no policy has been set, default to ignoring the file (no action taken).
+                            # Check if a policy has been set for the specific datatype or scope
+                            # or if there are any wildcard policies regarding the scope.
+                            # Datatype action has priority over scope action.
+                            # Specific scope has priority over wildcard scope.
+                            if "datatype" in i:
+                                if file_metadata_datatype in i["datatype"]:
+                                    try:
+                                        action = i["datatype_action"]
+                                        in_file_metadata_datatype = True
+                                        break
+                                    except:
+                                        logger(logging.WARNING, "RSE: %s, replica name %s, surl %s: Datatype %s found in the policy library, but an error occured while retrieving the policy assigned to it. No action will be taken.",
+                                               rse_key, replica_key, recoverable_replicas[vo][rse_key][replica_key]['surl'], file_metadata_datatype)
+                                        break
+                            if ("scope" in i) and not in_file_metadata_datatype:
```

Comment:
The action should be taken only if `in_file_metadata_datatype` if `datatype` is defined in the policy and `in_file_scope` if `scope` is defined in the policy. Therefore I would remove `and not in_file_metadata_datatype`

Reply:
The "not" clause is what makes sure that a datatype action has priority over a scope action. I'll remain the variable to make it a bit clearer.